File: 백준/Gold/3425. 고스택/고스택.cc
#include <bits/stdc++.h>

#define MAX 1000000000

using namespace std;

stack<long long> st;
vector<string> commands;
vector<long long> nums;
int N;

bool POP() {
	if (st.empty()) {
		return true;
	}
	st.pop();
	return false;
}

bool INV() {
	if (st.empty()) {
		return true;
	}
	long long top = st.top();
	st.pop();
	st.push(-top);
	return false;
}

bool DUP() {
	if (st.empty()) {
		return true;
	}
	st.push(st.top());
	return false;
}

bool SWP() {
	if (st.empty()) {
		return true;
	}
	long long first = st.top();
	st.pop();
	if (st.empty()) {
		return true;
	}
	long long second = st.top();
	st.pop();
	st.push(first);
	st.push(second);
	return false;
}

bool ADD() {
	if (st.empty()) {
		return true;
	}
	long long first = st.top();
	st.pop();
	if (st.empty()) {
		return true;
	}
	long long second = st.top();
	st.pop();
	st.push(first + second);
	return false;
}

bool SUB() {
	if (st.empty()) {
		return true;
	}
	long long first = st.top();
	st.pop();
	if (st.empty()) {
		return true;
	}
	long long second = st.top();
	st.pop();
	st.push(second - first);
	return false;
}

bool MUL() {
	if (st.empty()) {
		return true;
	}
	long long first = st.top();
	st.pop();
	if (st.empty()) {
		return true;
	}
	long long second = st.top();
	st.pop();
	st.push(first * second);
	return false;
}

bool DIV() {
	if (st.empty()) {
		return true;
	}
	long long first = st.top();
	st.pop();
	if (st.empty() || first == 0) {
		return true;
	}
	long long second = st.top();
	st.pop();
	long long div = abs(second) / abs(first);
	if (second < 0) {
		div = -div;
	}
	if (first < 0) {
		div = -div;
	}
	st.push(div);
	return false;
}

bool MOD() {
	if (st.empty()) {
		return true;
	}
	long long first = st.top();
	st.pop();
	if (st.empty() || first == 0) {
		return true;
	}
	long long second = st.top();
	st.pop();
	long long mod = abs(second) % abs(first);
	if (second < 0) {
		mod = -mod;
	}
	st.push(mod);
	return false;
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	cout.tie(nullptr);
	string command;
	long long num;
	bool error;
	int index;
	while (true) {
		while (true) {
			cin >> command;
			if (command == "NUM") {
				cin >> num;
				nums.emplace_back(num);
			} else if (command == "END") {
				break;
			} else if (command == "QUIT") {
				return 0;
			}
			commands.emplace_back(command);
		}
		cin >> N;

		for (int i = 0; i < N; i++) {
			error = false;
			index = 0;
			int first;
			cin >> first;
			st.push(first);
			for (const auto &item: commands) {
				if (item == "NUM") {
					st.push(nums[index++]);
				} else if (item == "POP") {
					error = POP();
				} else if (item == "INV") {
					error = INV();
				} else if (item == "DUP") {
					error = DUP();
				} else if (item == "SWP") {
					error = SWP();
				} else if (item == "ADD") {
					error = ADD();
				} else if (item == "SUB") {
					error = SUB();
				} else if (item == "MUL") {
					error = MUL();
				} else if (item == "DIV") {
					error = DIV();
				} else if (item == "MOD") {
					error = MOD();
				}
				if (!st.empty() && (st.top() > MAX || st.top() < -MAX)) {
					error = true;
				}
				if (error) break;
			}
			if (error || st.size() != 1) {
				printf("ERROR\n");
			} else {
				printf("%lld\n", st.top());
			}
			while (!st.empty()) {
				st.pop();
			}
		}
		vector<long long>().swap(nums);
		vector<string>().swap(commands);
		printf("\n");
	}
}

File: 백준/Gold/3425. 고스택/고스택.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});
const input = [];

rl.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  const machines = input
    .join('\n')
    .split('\n\n')
    .map((v) => v.split('\n'))
    .slice(0, -1);

  solution(machines);
});

const commandMap = {
  NUM: (stack, number) => {
    stack.push(Number(number));
    return true;
  },
  POP: (stack) => {
    if (stack.length === 0) return false;
    stack.pop();
    return true;
  },
  INV: (stack) => {
    if (stack.length === 0) return false;
    stack[stack.length - 1] *= -1;
    return true;
  },
  DUP: (stack) => {
    if (stack.length === 0) return false;
    stack.push(stack[stack.length - 1]);
    return true;
  },
  SWP: (stack) => {
    if (stack.length < 2) return false;
    [stack[stack.length - 1], stack[stack.length - 2]] = [
      stack[stack.length - 2],
      stack[stack.length - 1],
    ];
    return true;
  },
  ADD: (stack) => {
    if (stack.length < 2) return false;
    const a = stack.pop();
    const b = stack.pop();
    stack.push(a + b);
    return true;
  },
  SUB: (stack) => {
    if (stack.length < 2) return false;
    const a = stack.pop();
    const b = stack.pop();
    stack.push(b - a);
    return true;
  },
  MUL: (stack) => {
    if (stack.length < 2) return false;
    const a = stack.pop();
    const b = stack.pop();
    stack.push(a * b);
    return true;
  },
  DIV: (stack) => {
    if (stack.length < 2) return false;
    const a = stack.pop();
    const b = stack.pop();
    if (a === 0) return false;
    let result = Math.trunc(Math.abs(b) / Math.abs(a));
    if (a * b < 0) result *= -1;
    stack.push(result);
    return true;
  },
  MOD: (stack) => {
    if (stack.length < 2) return false;
    const a = stack.pop();
    const b = stack.pop();
    if (a === 0) return false;
    let result = Math.abs(b) % Math.abs(a);
    if (b < 0) result *= -1;
    stack.push(result);
    return true;
  },
};

const MAX = Math.pow(10, 9);

const solution = (machines) => {
  const result = [];
  for (const machine of machines) {
    const [commands, [, ...inputs]] = machine
      .join('\n')
      .split('END\n')
      .map((v) => v.split('\n'));

    for (const number of inputs) {
      let error = false;
      const stack = [Number(number)];

      for (const command of commands) {
        if (!command) continue;
        const [op, number] = command.split(' ');
        if (!commandMap[op](stack, number)) {
          error = true;
          break;
        }
      }

      if (error || stack.length !== 1 || stack[0] > MAX || stack[0] < -MAX) {
        result.push('ERROR');
      } else {
        result.push(stack[0]);
      }
    }
    result.push('');
  }
  console.log(result.join('\n'));
};
